aten/src/ATen/native/mkldnn/xpu/detail/MatmulHelpers.h

```diff
@@ -0,0 +1,231 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+#include <ATen/native/mkldnn/xpu/detail/Attr.h>
+#include <ATen/native/mkldnn/xpu/detail/Utils.h>
+
+#include <oneapi/dnnl/dnnl.hpp>
+
+namespace at::native::onednn {
+
+class BiasHandler {
+  // This class handles the bias tensor for matmul operations.
+  // It checks the dimensions of the bias tensor and expands it if necessary
+  // for broadcasting.
+ public:
+  BiasHandler() = delete;
+  BiasHandler(
+      at::Tensor& bias,
+      int64_t mb,
+      int64_t m,
+      int64_t k,
+      int64_t n,
+      int64_t gemm_dim)
+      : b(bias),
+        mb(mb),
+        m(m),
+        k(k),
+        n(n),
+        gemm_dim(gemm_dim),
+        with_bias(bias.defined()) {}
+
+  void handle0D() {
+    TORCH_CHECK(
+        b.numel() == 1, "matmul supports 1 numel when bias dim is [] ...");
+    if (gemm_dim == 3) {
+      b = b.expand({mb, m, n}).contiguous();
+    } else {
+      b = b.expand({1, n}).contiguous();
+    }
+  }
+
+  void handle1D() {
+    TORCH_CHECK(
+        b.size(0) == n || b.size(0) == 1,
+        "matmul supports [n] or [1] when bias dim is 1...");
+    if (b.size(0) == 0) {
+      with_bias = false;
+    } else if (gemm_dim == 3) {
+      b = b.expand({mb, m, n}).contiguous();
+    } else if (gemm_dim == 2) {
+      b = b.expand({1, n}).contiguous();
+    }
+  }
+
+  void handle2D() {
+    TORCH_CHECK(
+        (b.size(0) == m && b.size(1) == n) ||
+            (b.size(0) == 1 && b.size(1) == n) ||
+            (b.size(0) == m && b.size(1) == 1) ||
+            (b.size(0) == 1 && b.size(1) == 1),
+        "matmul supports [m, n] or [1, n] or [m, 1] or [1, 1] when bias dim is 2 ...");
+    if (b.size(0) == 1 && b.size(1) == 1)
+      b = b.expand({1, n}).contiguous();
+  }
+
+  void handle3D() {
+    TORCH_CHECK(
+        at::are_expandable({mb, m, n}, b.sizes()),
+        "matmul bias must be expandable to:",
+        "{mb, m, n} where mb is the batch size, m is the number of rows, and n is the number of columns.",
+        " but got:",
+        b.sizes());
+    b = b.expand({mb, m, n}).contiguous();
+  }
+
+  bool is_with_bias() {
+    return with_bias;
+  }
+
+  void handle() {
+    with_bias = b.defined();
+    if (!with_bias)
+      return;
+    using HandlerFn = void (BiasHandler::*)();
+    std::unordered_map<int, HandlerFn> handler_map = {
```

Comment:
```suggestion
    static std::unordered_map<int, HandlerFn> handler_map = {
```
Avoid unordered_map constructor overhead each time.